traversal/scanner/GraphIterator.java

```diff
@@ -206,28 +206,51 @@ private void initialiseStart() {
     private void initialiseEnds() {
         assert toTraverse.isEmpty();
         toRevisit.addAll(procedure.endVertices());
-        direction = Direction.REVISIT;
+        direction = Direction.REVISIT_RETRIEVED;
     }
 
     private boolean computeAnswer() {
         while ((direction == Direction.TRAVERSE && !toTraverse.isEmpty()) ||
-                (direction == Direction.REVISIT && !toRevisit.isEmpty())) {
+                ((direction == Direction.REVISIT_ALL || direction == Direction.REVISIT_RETRIEVED) && !toRevisit.isEmpty())) {
             ProcedureVertex<?, ?> vertex;
             if (direction == Direction.TRAVERSE) {
                 toTraverse.remove(vertex = toTraverse.first());
                 traverse(vertex);
-            } else {
+            } else if (direction == Direction.REVISIT_ALL) {
                 toRevisit.remove(vertex = toRevisit.last());
                 revisit(vertex);
+            } else {
+                toRevisit.remove(vertex = toRevisit.last());
+                revisitRetrieved(vertex);
             }
         }
         return direction == Direction.TRAVERSE;
     }
 
     private void traverse(ProcedureVertex<?, ?> procedureVertex) {
         VertexTraverser vertexTraverser = vertexTraversers.get(procedureVertex);
-        if (vertexTraverser.findNextVertex()) success(vertexTraverser);
-        else failed(vertexTraverser);
+        if (vertexTraverser.findNextVertex()) {
+            success(vertexTraverser);
+            assert direction == Direction.TRAVERSE;
+        }
+        else {
+            failed(vertexTraverser);
+            direction = Direction.REVISIT_ALL;
+        }
+    }
+
+    private void revisit(ProcedureVertex<?, ?> procedureVertex) {
+        toTraverse.add(procedureVertex);
+        direction = Direction.TRAVERSE;
+    }
+
+    private void revisitRetrieved(ProcedureVertex<?, ?> vertex) {
```

Comment:
retrieved vertex revisiting redirects into regular `revisit` if the variable is retrieved, otherwise immediately fails the vertex search for the vertex and backtracks